checkphish_connector.py

```diff
@@ -0,0 +1,371 @@
+            r.status_code, r.text.replace("{", "{{").replace("}", "}}")
+        )
+
+        return RetVal(action_result.set_status(phantom.APP_ERROR, message), None)
+
+    def _process_response(self, r, action_result):
+        # store the r_text in debug data, it will get dumped in the logs if the action fails
+        if hasattr(action_result, "add_debug_data"):
+            action_result.add_debug_data({"r_status_code": r.status_code})
+            action_result.add_debug_data({"r_text": r.text})
+            action_result.add_debug_data({"r_headers": r.headers})
+
+        # Process each 'Content-Type' of response separately
+
+        # Process a json response
+        if "json" in r.headers.get("Content-Type", ""):
+            return self._process_json_response(r, action_result)
+
+        # Process an HTML response, Do this no matter what the api talks.
+        # There is a high chance of a PROXY in between phantom and the rest of
+        # world, in case of errors, PROXY's return HTML, this function parses
+        # the error and adds it to the action_result.
+        if "html" in r.headers.get("Content-Type", ""):
+            return self._process_html_response(r, action_result)
+
+        # it's not content-type that is to be parsed, handle an empty response
+        if not r.text:
+            return self._process_empty_response(r, action_result)
+
+        # everything else is actually an error at this point
+        message = "Can't process response from server. Status Code: {0} Data from server: {1}".format(
+            r.status_code, r.text.replace("{", "{{").replace("}", "}}")
+        )
+
+        return RetVal(action_result.set_status(phantom.APP_ERROR, message), None)
+
+    def _make_rest_call(self, endpoint, action_result, method="get", **kwargs):
+        # **kwargs can be any additional parameters that requests.request accepts
+
+        config = self.get_config()
+
+        resp_json = None
+
+        try:
+            request_func = getattr(requests, method)
+        except AttributeError:
+            return RetVal(
+                action_result.set_status(
+                    phantom.APP_ERROR, "Invalid method: {0}".format(method)
+                ),
+                resp_json,
+            )
+
+        # Create a URL to connect to
+        url = "{}{}".format(self._api_url, endpoint)
+
+        try:
+            r = request_func(
+                url, verify=config.get("verify_server_cert", False), **kwargs
+            )
+        except Exception as e:
+            return RetVal(
+                action_result.set_status(
+                    phantom.APP_ERROR,
+                    "Error Connecting to server. Details: {0}".format(self._get_error_message_from_exception(e)),
+                ),
+                resp_json,
+            )
+
+        return self._process_response(r, action_result)
+
+    def _handle_test_connectivity(self, param):
```

Comment:
I saw that the readme says this app does not implement test_connectivity. Why not?

Reply:
As this is a third-party community app, this is how the contributor has kept it. We don’t make changes in the community apps as we don't have a positive asset for that.

Reply:
Ah understood.  I do not know the history, but I think in general we should push back on community submission to include test_connectivity as it is vital for most customers (asset health page, for example). 

If this has been raised and approved, though, then I will approve the PR

Reply:
Hi @dfederschmidt , can you please confirm whether you are planning to implement the test-connectivity action or not?

Reply:
@joshquerinphantom the original omission of the `test connectivity` action is outlined in the original MR description.

> API does not offer suitable endpoint for connectivity testing (only offers to detonate a URL, which would reduce quota)

The API does not offer an authenticated endpoint that we can call other than the one called for `detonate url`, reducing quota. I suppose we can call the endpoint for a well-known URL (eg. `google.com` or `checkphish.ai`) and document that running the test connectivity action will reduce quota.

I'll push that addition in a bit.

Reply:
Please review https://github.com/splunk-soar-connectors/checkphish/pull/1/commits/9ffa524314c7dd972b9330eb2fcad04a8dd422cd

Reply:
@dfederschmidt Reviewed and verified.